File: apps/mobile/lib/core/logic/fabrary_match.dart
import '../data/card_repository.dart';
import '../models/card_model.dart';

const _artTreatments = [
  'extended art',
  'full art',
  'alternate art',
  'alternate text',
  'alternate border',
  'alt art',
  'marvel',
  'treasure',
];

const _treatmentAliases = {
  'full art': ['full art', 'marvel', 'treasure'],
  'alternate art': ['alternate art', 'alt art'],
  'alt art': ['alternate art', 'alt art'],
  'extended art': ['extended art'],
  'alternate border': ['alternate border'],
  'alternate text': ['alternate text'],
};

final _foilNumberSuffix = RegExp(r'(cf|rf|gf)$');

const _pitchByName = {'red': '1', 'yellow': '2', 'blue': '3'};

class FabraryUnmatched {
  const FabraryUnmatched({
    required this.name,
    required this.setNumber,
    required this.foiling,
    required this.treatment,
    required this.edition,
  });

  final String name;
  final String setNumber;
  final String foiling;
  final String treatment;
  final String edition;

  Map<String, String> toJson() => {
        'name': name,
        'setNumber': setNumber,
        'foiling': foiling,
        'treatment': treatment,
        'edition': edition,
      };
}

class FabraryMatch {
  const FabraryMatch.hit(String id)
      : printingId = id,
        unmatched = null;

  const FabraryMatch.miss(FabraryUnmatched miss)
      : printingId = null,
        unmatched = miss;

  final String? printingId;
  final FabraryUnmatched? unmatched;

  Map<String, dynamic> toJson() {
    if (unmatched != null) return {'unmatched': unmatched!.toJson()};
    return {'printingId': printingId};
  }
}

String _rowValue(Map<String, dynamic> row, String key, [String alt = '']) {
  final value = row[key] ?? (alt.isEmpty ? null : row[alt]);
  return value == null ? '' : '$value';
}

String printingIdOf(dynamic card) {
  if (card is CardModel) return card.id;
  if (card is Map) {
    return '${card['id'] ?? card['_uniqueId'] ?? ''}';
  }
  return '';
}

String collectorNumberOf(dynamic card) {
  if (card is CardModel) return card.collectorNumber ?? '';
  if (card is Map) {
    return '${card['collectorNumber'] ?? card['extNumber'] ?? card['collector_number'] ?? ''}';
  }
  return '';
}

String subTypeOf(dynamic card) {
  if (card is CardModel) return card.subTypeName ?? '';
  if (card is Map) {
    return '${card['subTypeName'] ?? card['sub_type_name'] ?? ''}';
  }
  return '';
}

String setNameOf(dynamic card) {
  if (card is CardModel) return card.setName ?? '';
  if (card is Map) {
    return '${card['setName'] ?? card['_setName'] ?? card['set_name'] ?? ''}';
  }
  return '';
}

String nameOf(dynamic card) {
  if (card is CardModel) return card.name;
  if (card is Map) return '${card['name'] ?? ''}';
  return '';
}

String pitchOf(dynamic card) {
  if (card is CardModel) return card.pitch ?? '';
  if (card is Map) {
    final raw = card['pitch'] ?? card['extPitchValue'];
    if (raw == null || raw == '') return '';
    return '$raw';
  }
  return '';
}

bool _hasArtTreatment(String name) {
  final lower = name.toLowerCase();
  return _artTreatments.any((token) => lower.contains('($token)'));
}

bool _isRegularPrinting(dynamic card) => !_hasArtTreatment(nameOf(card));

String _editionHaystack(dynamic card) =>
    '${subTypeOf(card)} ${setNameOf(card)}'.toLowerCase();

bool _hasPitchColorName(String name) {
  final lower = name.toLowerCase();
  return lower.contains('(red)') ||
      lower.contains('(yellow)') ||
      lower.contains('(blue)');
}

bool _matchesPitch(dynamic card, String fabraryPitch) {
  final want = fabraryPitch.trim().toLowerCase();
  if (want.isEmpty) return true;
  final mapped = _pitchByName[want];
  final catalogPitch = pitchOf(card);
  final hasNumericPitch = catalogPitch.isNotEmpty && catalogPitch != '0';
  if (mapped != null && hasNumericPitch && catalogPitch == mapped) return true;
  final name = nameOf(card).toLowerCase();
  if (name.contains('($want)')) return true;
  if (!hasNumericPitch && !_hasPitchColorName(name)) return true;
  return false;
}

bool _matchesFoil(dynamic card, String foiling) {
  final want = foiling.trim().toLowerCase();
  final sub = subTypeOf(card).toLowerCase();
  if (want.isEmpty) {
    return !RegExp(r'\brainbow\b').hasMatch(sub) &&
        !RegExp(r'\bcold\b').hasMatch(sub) &&
        !RegExp(r'\bgold\b').hasMatch(sub);
  }
  if (want.contains('rainbow')) return sub.contains('rainbow');
  if (want.contains('cold')) return sub.contains('cold');
  if (want.contains('gold')) return sub.contains('gold');
  return sub.contains(want);
}

List<String> _treatmentNeedles(String treatment) {
  final want = treatment.trim().toLowerCase();
  if (want.isEmpty) return const <String>[];
  return _treatmentAliases[want] ?? <String>[want];
}

bool _nameHasTreatment(String name, List<String> needles) {
  final lower = name.toLowerCase();
  return needles.any((token) => lower.contains('($token)'));
}

List<dynamic> _applyTreatmentFilter(List<dynamic> candidates, String treatment) {
  final want = treatment.trim();
  if (want.isEmpty) {
    final regular =
        candidates.where((card) => !_hasArtTreatment(nameOf(card))).toList();
    return regular.isNotEmpty ? regular : candidates;
  }
  final needles = _treatmentNeedles(want);
  final exact = candidates
      .where((card) => _nameHasTreatment(nameOf(card), needles))
      .toList();
  if (exact.isNotEmpty) return exact;
  final artish =
      candidates.where((card) => _hasArtTreatment(nameOf(card))).toList();
  return artish.isNotEmpty ? artish : candidates;
}

List<dynamic> _applyFoilFilter(List<dynamic> candidates, String foiling) {
  final matched =
      candidates.where((card) => _matchesFoil(card, foiling)).toList();
  if (matched.isNotEmpty) return matched;
  if (foiling.trim().isEmpty) return matched;
  return candidates.where((card) => subTypeOf(card).trim().isEmpty).toList();
}

final _collectorNumberSplit = RegExp(r'\s*//\s*|\s*/\s*');

List<String> collectorNumberKeys(String? raw) {
  final text = raw ?? '';
  if (text.trim().isEmpty) return const <String>[];
  final keys = <String>[];
  final seen = <String>{};
  void add(String value) {
    final key = collectorNumberKey(value);
    if (key == null || !seen.add(key)) return;
    keys.add(key);
    final stripped = key.replaceFirst(_foilNumberSuffix, '');
    if (stripped != key && stripped.length >= 5 && seen.add(stripped)) {
      keys.add(stripped);
    }
  }

  add(text);
  for (final part in text.split(_collectorNumberSplit)) {
    final trimmed = part.trim();
    if (trimmed.isNotEmpty) add(trimmed);
  }
  return keys;
}

bool _hasEditionToken(dynamic card, String token) {
  final text = _editionHaystack(card);
  if (token == 'first') {
    return RegExp(r'\b1st\b').hasMatch(text) ||
        RegExp(r'\bfirst\b').hasMatch(text);
  }
  if (token == 'unlimited') return RegExp(r'\bunlimited\b').hasMatch(text);
  if (token == 'alpha') return RegExp(r'\balpha\b').hasMatch(text);
  return false;
}

bool _matchesEdition(dynamic card, String edition) {
  final want = edition.trim().toLowerCase();
  if (want.isEmpty) {
    return !_hasEditionToken(card, 'first') &&
        !_hasEditionToken(card, 'unlimited') &&
        !_hasEditionToken(card, 'alpha');
  }
  if (want == 'first') return _hasEditionToken(card, 'first');
  if (want == 'unlimited') return _hasEditionToken(card, 'unlimited');
  if (want == 'alpha') return _hasEditionToken(card, 'alpha');
  return _editionHaystack(card).contains(want);
}

FabraryUnmatched _unmatchedOf(Map<String, dynamic> row) => FabraryUnmatched(
      name: _rowValue(row, 'Name', 'name'),
      setNumber: _rowValue(row, 'Set number', 'setNumber'),
      foiling: _rowValue(row, 'Foiling', 'foiling'),
      treatment: _rowValue(row, 'Treatment', 'treatment'),
      edition: _rowValue(row, 'Edition', 'edition'),
    );

dynamic _pickCandidate(List<dynamic> candidates) {
  if (candidates.isEmpty) return null;
  if (candidates.length == 1) return candidates.first;
  final regular = candidates.where(_isRegularPrinting).toList();
  final pool = regular.isNotEmpty ? regular : candidates;
  pool.sort((a, b) {
    final byName = nameOf(a).compareTo(nameOf(b));
    if (byName != 0) return byName;
    final byNumber = collectorNumberOf(a).compareTo(collectorNumberOf(b));
    if (byNumber != 0) return byNumber;
    return subTypeOf(a).compareTo(subTypeOf(b));
  });
  return pool.first;
}

Map<String, List<dynamic>> buildFabrarySetIndex(Iterable<dynamic> catalog) {
  final index = <String, List<dynamic>>{};
  for (final card in catalog) {
    for (final key in collectorNumberKeys(collectorNumberOf(card))) {
      final list = index.putIfAbsent(key, () => <dynamic>[]);
      if (!list.contains(card)) list.add(card);
    }
  }
  return index;
}

/// Match one Fabrary owned row to a catalog Printing.
///
/// [catalog] may be [CardModel]s, catalog-shaped maps, or a prebuilt set-code
/// index from [buildFabrarySetIndex]. Identifier is ignored.
FabraryMatch matchFabraryRow(
  Map<String, dynamic> row,
  dynamic catalog,
) {
  final setNumber = _rowValue(row, 'Set number', 'setNumber');
  final key = collectorNumberKey(setNumber);
  if (key == null) return FabraryMatch.miss(_unmatchedOf(row));

  final index = catalog is Map<String, List<dynamic>>
      ? catalog
      : buildFabrarySetIndex(
          catalog is Iterable ? catalog as Iterable<dynamic> : const <dynamic>[],
        );
  final pool = index[key] ?? const <dynamic>[];
  final pitched = pool
      .where(
        (card) =>
            _matchesPitch(card, _rowValue(row, 'Pitch', 'pitch')) &&
            _matchesEdition(card, _rowValue(row, 'Edition', 'edition')),
      )
      .toList();
  final foiled =
      _applyFoilFilter(pitched, _rowValue(row, 'Foiling', 'foiling'));
  final candidates =
      _applyTreatmentFilter(foiled, _rowValue(row, 'Treatment', 'treatment'));

  final chosen = _pickCandidate(candidates);
  if (chosen == null) return FabraryMatch.miss(_unmatchedOf(row));
  return FabraryMatch.hit(printingIdOf(chosen));
}

File: apps/web/src/utils/fabraryMatch.js
/**
 * Fabrary owned row → catalog Printing. Identifier is not a match key.
 *
 * Filter order is pinned by packages/contracts/fabrary_printing_match.json.
 */

const ART_TREATMENTS = [
    'extended art',
    'full art',
    'alternate art',
    'alternate text',
    'alternate border',
    'alt art',
    'marvel',
    'treasure',
];

/** Fabrary Treatment labels → catalog name parentheticals. */
const TREATMENT_ALIASES = {
    'full art': ['full art', 'marvel', 'treasure'],
    'alternate art': ['alternate art', 'alt art'],
    'alt art': ['alternate art', 'alt art'],
    'extended art': ['extended art'],
    'alternate border': ['alternate border'],
    'alternate text': ['alternate text'],
};

const FOIL_NUMBER_SUFFIX = /(cf|rf|gf)$/;

const PITCH_BY_NAME = { red: '1', yellow: '2', blue: '3' };
const NAME_BY_PITCH = { 1: 'red', 2: 'yellow', 3: 'blue' };

export function collectorNumberKey(raw) {
    if (raw == null) return null;
    const key = String(raw).toLowerCase().replace(/[^a-z0-9]+/g, '');
    return key === '' ? null : key;
}

/** Index keys for a collector number, including `//` / `/` faces and foil suffixes. */
export function collectorNumberKeys(raw) {
    const text = raw == null ? '' : String(raw);
    if (!text.trim()) return [];
    const segments = text.split(/\s*\/\/\s*|\s*\/\s*/).map((part) => part.trim()).filter(Boolean);
    const keys = [];
    const seen = new Set();
    const add = (value) => {
        const key = collectorNumberKey(value);
        if (!key || seen.has(key)) return;
        seen.add(key);
        keys.push(key);
        const stripped = key.replace(FOIL_NUMBER_SUFFIX, '');
        if (stripped !== key && stripped.length >= 5 && !seen.has(stripped)) {
            seen.add(stripped);
            keys.push(stripped);
        }
    };
    add(text);
    for (const part of segments) add(part);
    return keys;
}

export function printingIdOf(card) {
    return card?.id || card?._uniqueId || '';
}

export function collectorNumberOf(card) {
    return card?.collectorNumber || card?.extNumber || card?.collector_number || '';
}

export function subTypeOf(card) {
    return card?.subTypeName || card?.sub_type_name || '';
}

export function setNameOf(card) {
    return card?.setName || card?._setName || card?.set_name || '';
}

export function pitchOf(card) {
    const raw = card?.pitch ?? card?.extPitchValue;
    if (raw === undefined || raw === null || raw === '') return '';
    return String(raw);
}

function stripNameSetCode(name) {
    const trimmed = String(name || '').trim();
    const stripped = trimmed.replace(/\s+-\s+[A-Z0-9]{2,}\s*$/i, '').trim();
    return stripped === '' ? trimmed : stripped;
}

export function nameQualifier(name) {
    const match = /\(([^)]*)\)\s*$/.exec(stripNameSetCode(name));
    return match ? match[1] : null;
}

function editionHaystack(card) {
    return `${subTypeOf(card)} ${setNameOf(card)}`.toLowerCase();
}

function hasArtTreatment(name) {
    const lower = String(name || '').toLowerCase();
    return ART_TREATMENTS.some((token) => lower.includes(`(${token})`));
}

function isRegularPrinting(card) {
    return !hasArtTreatment(card?.name || '');
}

function hasPitchColorName(name) {
    const lower = String(name || '').toLowerCase();
    return lower.includes('(red)') || lower.includes('(yellow)') || lower.includes('(blue)');
}

function matchesPitch(card, fabraryPitch) {
    const want = String(fabraryPitch || '').trim().toLowerCase();
    if (!want) {
        // Blank pitch: do not require a color, but prefer colorless when
        // later tie-breaking. Still accept any remaining candidate.
        return true;
    }
    const mapped = PITCH_BY_NAME[want];
    const catalogPitch = pitchOf(card);
    const hasNumericPitch = catalogPitch && catalogPitch !== '0';
    if (mapped && hasNumericPitch && String(catalogPitch) === mapped) return true;
    const name = String(card?.name || '').toLowerCase();
    if (name.includes(`(${want})`)) return true;
    if (mapped && NAME_BY_PITCH[catalogPitch] === want) return true;
    // Super Slam-style names omit (Red)/(Blue); collector number already
    // identifies the pitch when the catalog row has no pitch field.
    if (!hasNumericPitch && !hasPitchColorName(name)) return true;
    return false;
}

function matchesFoil(card, foiling) {
    const want = String(foiling || '').trim().toLowerCase();
    const sub = subTypeOf(card).toLowerCase();
    if (!want) {
        return !/\brainbow\b/.test(sub) && !/\bcold\b/.test(sub) && !/\bgold\b/.test(sub);
    }
    if (want.includes('rainbow')) return sub.includes('rainbow');
    if (want.includes('cold')) return sub.includes('cold');
    if (want.includes('gold')) return sub.includes('gold');
    return sub.includes(want);
}

function treatmentNeedles(treatment) {
    const want = String(treatment || '').trim().toLowerCase();
    if (!want) return [];
    return TREATMENT_ALIASES[want] || [want];
}

function nameHasTreatment(name, needles) {
    const lower = String(name || '').toLowerCase();
    return needles.some((token) => lower.includes(`(${token})`));
}

function applyTreatmentFilter(candidates, treatment) {
    const want = String(treatment || '').trim();
    if (!want) {
        const regular = candidates.filter((card) => !hasArtTreatment(card?.name || ''));
        return regular.length > 0 ? regular : candidates;
    }
    const needles = treatmentNeedles(want);
    const exact = candidates.filter((card) => nameHasTreatment(card?.name || '', needles));
    if (exact.length > 0) return exact;
    // GEM / promo / Marvel rows often omit Fabrary's exact treatment words.
    // Prefer any remaining art-treated printing so we do not land on the
    // ordinary version when a special version exists (FR-010).
    const artish = candidates.filter((card) => hasArtTreatment(card?.name || ''));
    return artish.length > 0 ? artish : candidates;
}

function applyFoilFilter(candidates, foiling) {
    const matched = candidates.filter((card) => matchesFoil(card, foiling));
    if (matched.length > 0) return matched;
    const want = String(foiling || '').trim();
    if (!want) return matched;
    return candidates.filter((card) => !String(subTypeOf(card) || '').trim());
}

function hasEditionToken(card, token) {
    const text = editionHaystack(card);
    if (token === 'first') return /\b1st\b/.test(text) || /\bfirst\b/.test(text);
    if (token === 'unlimited') return /\bunlimited\b/.test(text);
    if (token === 'alpha') return /\balpha\b/.test(text);
    return false;
}

function matchesEdition(card, edition) {
    const want = String(edition || '').trim().toLowerCase();
    if (!want) {
        return !hasEditionToken(card, 'first')
            && !hasEditionToken(card, 'unlimited')
            && !hasEditionToken(card, 'alpha');
    }
    if (want === 'first') return hasEditionToken(card, 'first');
    if (want === 'unlimited') return hasEditionToken(card, 'unlimited');
    if (want === 'alpha') return hasEditionToken(card, 'alpha');
    return editionHaystack(card).includes(want);
}

export function buildSetCodeIndex(catalog) {
    const index = new Map();
    for (const card of catalog || []) {
        for (const key of collectorNumberKeys(collectorNumberOf(card))) {
            const list = index.get(key);
            if (list) {
                if (!list.includes(card)) list.push(card);
            } else {
                index.set(key, [card]);
            }
        }
    }
    return index;
}

function unmatchedOf(row) {
    return {
        unmatched: {
            name: String(row?.Name ?? row?.name ?? ''),
            setNumber: String(row?.['Set number'] ?? row?.setNumber ?? ''),
            foiling: String(row?.Foiling ?? row?.foiling ?? ''),
            treatment: String(row?.Treatment ?? row?.treatment ?? ''),
            edition: String(row?.Edition ?? row?.edition ?? ''),
        },
    };
}

function pickCandidate(candidates) {
    if (candidates.length === 0) return null;
    if (candidates.length === 1) return candidates[0];
    const regular = candidates.filter(isRegularPrinting);
    const pool = regular.length > 0 ? regular : candidates;
    const sorted = [...pool].sort((a, b) => {
        const byName = String(a?.name || '').localeCompare(String(b?.name || ''));
        if (byName !== 0) return byName;
        const byNumber = String(collectorNumberOf(a)).localeCompare(String(collectorNumberOf(b)));
        if (byNumber !== 0) return byNumber;
        return String(subTypeOf(a)).localeCompare(String(subTypeOf(b)));
    });
    return sorted[0];
}

/**
 * @param {Object} row Fabrary owned row
 * @param {Object[]} catalog catalog-shaped printings
 * @returns {{ printingId: string } | { unmatched: Object }}
 */
export function matchFabraryRow(row, catalog) {
    const setNumber = row?.['Set number'] ?? row?.setNumber ?? '';
    const key = collectorNumberKey(setNumber);
    if (!key) return unmatchedOf(row);

    const index = Array.isArray(catalog) ? buildSetCodeIndex(catalog) : catalog;
    const pool = (index.get ? index.get(key) : null) || [];
    const pitched = pool.filter((card) =>
        matchesPitch(card, row?.Pitch ?? row?.pitch)
        && matchesEdition(card, row?.Edition ?? row?.edition),
    );
    const foiled = applyFoilFilter(pitched, row?.Foiling ?? row?.foiling);
    const candidates = applyTreatmentFilter(foiled, row?.Treatment ?? row?.treatment);

    const chosen = pickCandidate(candidates);
    if (!chosen) return unmatchedOf(row);
    return { printingId: printingIdOf(chosen) };
}
